### PR TITLE
Check sns topic permission at start of buildafi and warn user

### PR DESCRIPTION
**Pull request by rahulb10**
Rather than exiting with a unhandled exception when SNS-related permissions failures happen at the end of `buildafi`, check whether the topic exists (or can be created) early in `buildafi` and warn the user that the script will be unable to send email (and log the details of the exception) but continue to finish `buildafi` because failure in sending the notification probably should not be a fatal error for the manager.

This PR also introduces `pytest` driven unit tests for the firesim manager.  The tests make use of `moto` to mock the backend of `boto` and prevent tests from actually calling out to AWS API's.  They also utilize `unittest.mock` and `botocore.stub.Stubber` to inject desired testing stimulus to the code under test.

### CODE REVIEW

**Comment by timsnyder-siv** [comment]
@rahulb10 , you need to edit your PR so that it is going into `firesim:dev`, then you will see only one commit in the PR that you expect.

File: deploy/firesim
Comment (timsnyder-siv) [inline deploy/firesim:131]: 1. You should capture this in a new function in https://github.com/sifive/firesim/blob/Topic_permission_check/deploy/awstools/awstools.py and refactor the other two uses to also call the new function:
   * https://github.com/sifive/firesim/blob/fafb94de6926350aebe7cd1f4bea63fc840312a3/deploy/awstools/awstools.py#L349-L358
   * https://github.com/sifive/firesim/blob/fafb94de6926350aebe7cd1f4bea63fc840312a3/deploy/awstools/awstools.py#L375-L385
   
2. You should put this in a try/except block with a helpful error message sent to `rootLogger.critical` before an empty `raise` to rethrow the exception.

**Review by timsnyder-siv — CHANGES_REQUESTED** [review]
In addition to refactoring that code into a single function please make sure that your git name and email address is setup correctly so that your commits don't say that they were done by "Cloud User".

File: deploy/firesim
Comment (rahulb10) [reply deploy/firesim]: > refactor the other two uses to also call the new function
I wrapped the topic creation in a function.
> You should put this in a try/except block with a helpful error message
Created the same and to exit after the error.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:124]: Also, I think it would be better to only have the `client.create_topic` call inside of our new function in the `try` block. It will more precisely catch what we're trying to catch and also handle the permissions errors in other different uses of the new function that we may not think about today.

**Review by timsnyder-siv — CHANGES_REQUESTED** [review]
Getting better.  There are still some things I'd like to address.

File: deploy/firesim
Comment (rahulb10) [reply deploy/firesim:124]: > Also, I think it would be better to only have the `client.create_topic` call inside of our new function in the `try` block.

I didn't get this. We have to anyway get the 'snsname' right to execute 'client.create_topic' ?

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:365]:
```suggestion
            rootLogger.critical("You don't have permissions to perform \"Topic Creation \". Required to send you email notifications. Please contact your IT administrator")
        else:
            rootLogger.critical("Unknown exception is encountered while trying to perform \"Topic Creation\"")
        rootLogger.critical(err)
        assert False
```

Ugh, github's rendering of this patch is really ugly, should just be moving a line down with an indent change.  Anyway, move the `rootLogger.critical(err)` to just before the `assert False` so that it is logged either way.

Comment (rahulb10) [reply deploy/awstools/awstools.py]: Done

**Comment by timsnyder-siv** [comment]
I think this is yet another place where it would be really good to mock boto3 and add a unit test that checks for this new behavior of the new function `awstools.get_snsname_arn`.  Talking offline with @rahulb10 about this and one of us will add a test.

**Comment by timsnyder-siv** [comment]
@rahulb10 I pushed a couple of tests.  Figuring out all of the magic to properly mock boto and also prevent it from being able to reach out to AWS while testing took a bit longer than I though it would.

File: deploy/tests/awstools/test_awstools.py
Comment (timsnyder-siv) [inline deploy/tests/awstools/test_awstools.py:24]:
```suggestion
    # the boto3.client factory (in the second test below) and also because we want to be able to run these
```

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:367]:
```suggestion
    except client.exceptions.ClientError as err:
        if 'AuthorizationError' in repr(err): 
            rootLogger.warning("Will not be able to send you email notifications.  You do not have AWS IAM permission  \"sns:CreateTopic\". Please contact your administrator")
            rootLogger.debug(err)

        else:
            rootLogger.critical("Unknown exception is encountered while trying to perform \"Topic Creation\". Please notify the FireSim team.")
            rootLogger.critical(err)
            assert False
```

@rahulb10 I realized that somehow we've steered away from the original intent here.  It definitely shouldn't be a fatal error when we're unable to create the topic and we know it was for AuthError.  That would prevent people from being able to build without the notifications.  It's debatable whether the unknown exception should `assert False` but maybe it should to encourage folks to let us know about other failures we could handle more helpfully.

I'm realizing that we probably need to update the docs to describe this and also update the code used in `firesim managerinit` where it tries to create the topic and subscribe to it.  Subscription is a separate permission and could fail differently.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:377]: @rahulb10 this is the the call I think we should put into a another try/except block with similar structure to what we've done in `get_snsname_arn` except we will want to say they are missing "sns:Subscribe" permission.

Looking down here, if we make the `get_snsname_arn()` AuthError failure non fatal, we need to either reraise an exception that can be caught by the callers or return `None` so the callers don't make further AWS requests with a bogus arn.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [reply deploy/awstools/awstools.py:367]: If we change the AuthError to be non-fatal (like I think we should), we'll also need to update the test I added becuase it expects the AuthError to be fatal.

File: deploy/firesim
Comment (timsnyder-siv) [inline deploy/firesim:123]: This is a great improvement but we could still have the case where the user has `sns:CreateTopic` but they are missing the `sns:Publish` permission and thus still can't send the message.  It's different than the failure case @rahulb10 and I hit but it we should put try/catch blocks around the other AWS API calls

File: deploy/tests/awstools/test_awstools.py
Comment (timsnyder-siv) [inline deploy/tests/awstools/test_awstools.py:79]: I dislike this pattern of using `@patch` as a decorator because it is unclear that the `@patch` decorator is providing the parameter, especially when mixed with other decorators and given that pytest provides fixtures as parameters.  

There is a `pytest-mock` plugin to pytest that thinly wraps `mock` so that it's usage is a little cleaner in pytest.  if we use that plugin, this code would look like this
```python
    @mock_sns
    def test_get_snsname_arn_auth_exception_handling(self, mocker):
        """AuthorizationError is intercepted and re-raised as AssertionError"""
        # local imports of code-under-test ensure moto has mocks
        # registered before any possible calls out to AWS
        from awstools.awstools import get_snsname_arn
        
        mocker.patch('awstools.awstools.aws_resource_names', return_value={'snsname': 'Testing'})

        # create a mock SNS client that returns what we tell it to
        client = boto3.client('sns')
        stub = Stubber(client)
        stub.add_client_error('create_topic', service_error_code='AuthorizationError')
        stub.activate()


        # since firesim manager code doesn't take clients as method parameters
        # now we mock boto3.client to return our stubbed client
        mocked_client_factory = patch.object(boto3._get_default_session(), 'client', return_value=client)
        with pytest.raises(AssertionError):
              get_snsname_arn()

        # TODO we could mock rootLogger.critical to capture it's calls and args and validate that we're seeing the correct "nice" message

        # make sure get_snsname_arn() actually called out to get a sns
        # client, otherwise we aren't testing what we think we are
        mocked_client_factory.assert_called_once_with('sns')

        awstools.awstools.aws_resource_names.assert_called_once()
```

Or, we could do one of the alternatives discussed in https://github.com/pytest-dev/pytest-mock/blob/3a4b13b9413e9c67b91e55b972698da1241e3ee2/README.rst#why-bother-with-a-plugin 

I do think that the pytest-mock plugin makes the test code cleaner.  I didn't add it because:
* I was trying to limit the number of libraries I added 
* I was lazy and didn't want to spend time figuring out which version would work with the version of mock that we have installed.

File: deploy/tests/awstools/test_awstools.py
Comment (timsnyder-siv) [inline deploy/tests/awstools/test_awstools.py:62]: It's probably useful to add a comment here that says something like 
```
# we use botocore.stub.Stubber() here instead of mock.patch because boto dynamically
# creates classes at runtime for it's service-specific client objects.  mock.patch can only patch
# static classes that exist in the python symbol table.  botocore.stub.Stubber() is explicitly created for this
# see: https://github.com/boto/boto3/issues/2485
```

File: deploy/awstools/awstools.py
Comment (rahulb10) [reply deploy/awstools/awstools.py:377]: I have to still retain `assert False` in get_snsname_arn() if at all I use try/except block in `subscribe_to_firesim_topic()` . In that case for it to remain non-fatal I have to use try/except block in all the functions that call `get_snsname_arn()` including the check we are having at the beginning of buildafi.

File: deploy/firesim
Comment (timsnyder-siv) [reply deploy/firesim:123]: My concern here is addressed by the changes in b25cd16

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:412]:
```suggestion
            rootLogger.warning("You don't have permissions to publish to firesim notifications")
```
You should say publish here instead of subscribe.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:416]:
```suggestion
```
I would remove this logging statement.  It may have been helpful while you were developing but I don't think it adds information during normal usage.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:393]:
```suggestion
```
I would remove this logging statement.  It may have been helpful while you were developing but I don't think it adds information during normal usage.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [inline deploy/awstools/awstools.py:386]: printing this message should move into an `else:` clause after the `except:` so that it is only printed if an exception is not raised when trying to do `client.subscribe()`

**Comment by timsnyder-siv** [comment]
As we're adding all of these error handling try/except clauses around calls to `boto_client.<action>`, they look very similar and a future work might be good to either:
* create our own context-manager class that we can use with a simple `with HandleBotoClientErrors():` block around the action calls instead of having to repeat the except clauses many different places.
* create our own wrapped boto client class that we could put the error handling code in one place. The wrapper class might be more flexible.  Then when we want to handle errors in a standard way, we could do something like `client = ExceptionHandler(boto3.client(<service>))`.  

Like I said though, let's leave this for future work unless the BAR folks disagree.

File: deploy/awstools/awstools.py
Comment (timsnyder-siv) [reply deploy/awstools/awstools.py]: This is also resolved in b25cd16

File: deploy/awstools/awstools.py
Comment (rahulb10) [reply deploy/awstools/awstools.py:386]: If the message is in `try`:, it would only get printed if an exception is not raised right?

Comment (timsnyder-siv) [reply deploy/awstools/awstools.py:386]: Yes.  You're correct.  Since it is after the point where an exception would be raised it's correct the way it is.  The only advantage of moving it might be for a little clarity and so that there's no way it could raise an exception that would be made non-fatal.  I don't think that's a big enough deal to move it.

File: deploy/tests/awstools/test_awstools.py
Comment (timsnyder-siv) [inline deploy/tests/awstools/test_awstools.py:70]: You should probably also assert that the return value is `None`.

Comment (timsnyder-siv) [reply deploy/tests/awstools/test_awstools.py]: addressed in 2a29eb8

File: deploy/tests/awstools/test_awstools.py
Comment (timsnyder-siv) [reply deploy/tests/awstools/test_awstools.py]: addressed in c24b334

**Review by abejgonzalez — APPROVED** [review]
LGTM. Though, I think it would be nice to add the tests to the CI before this is merged in.

**Comment by timsnyder-siv** [comment]
> it would be nice to add the tests to the CI before this is merged in.
@davidbiancolin said he'd take care of adding them to the CI.  David, I'll leave it up to you whether you want to add them in this PR and merge them passing in CI or add them in a follow-on.

**Comment by davidbiancolin** [comment]
Yeah i'll add them now.

**Comment by davidbiancolin** [comment]
Don't forget to a changelog entry in the dev->master PR.

**Comment by timsnyder-siv** [comment]
@rahulb10 unless you have more tests to add now for the changes we made to the other methods in awstools, I think we should go ahead to "squash and merge" this PR.

**Pull request merged**
